### ISSUE TITLE
`[None] * int` not widened to match expected list type

### ISSUE THREAD

**Issue by Jeremiah-England** [astral-sh/ty#3002]
### Summary

See https://play.ty.dev/092bbf5e-a753-4e30-bd29-1a80fb86304d and https://github.com/astral-sh/ty/issues/2972#issuecomment-4023996886.

```python
from typing import reveal_type
x: list[int] = list()

y: list[int | None] = [None] * len(x)  # Object of type `list[None]` is not assignable to `list[int | None]` (invalid-assignment)
y: list[int | None] = [None]  # OK


def func() -> list[int | None]:
    if ...:
        return [None] * len(x) # Return type does not match returned value: expected `list[int | None]`, found `list[None]` (invalid-return-type)
    else:
        return [None]  # OK

def func(x: list[int | None]):
    ...

func([None] * len(x)) # Argument to function `func` is incorrect: Expected `list[int | None]`, found `list[None]` (invalid-argument-type)
func(reveal_type([None]))  # OK  (Revealed type: `list[int | None]`)
```

### Version

ty 0.0.21

**Comment by ibraheemdev** [comment]
This one is a little tricky because the method resolution for `list[None].__mul__` happens after type inference, so it's not exactly clear to me how the type context should be propagated to the LHS expression other than "try and see if we get a more assignable type"?

**Comment by AlexWaygood** [comment]
I don't have an immediate suggestion for how to fix it, but I do think we need to find _a_ solution here. Mypy, pyright, pyrefly and zuban all emit no errors on a slightly modified version of this snippet -- here's a gist you can load into [multiplay](https://github.com/astral-sh/multiplay): 51a54fb002b68fd176321ec029254883.

**Comment by carljm** [comment]
Interestingly, pyright/pyrefly/mypy all start erroring if you add a `reveal_type` wrapper like this:

```py
y: list[int | None] = reveal_type([None]) * len(x)
```

Which suggests they might be pattern-matching in a shallow way?

Zuban handles that gracefully -- might be worth looking into what it's doing here.

**Comment by charliermarsh** [comment]
Codex suggested this which (I guess) would align with Carl's comment on shallow matching:
```diff
diff --git a/crates/ty_python_semantic/src/types/infer/builder/binary_expressions.rs b/crates/ty_python_semantic/src/types/infer/builder/binary_expressions.rs
index f56cf80b39..aa5f5cdfdb 100644
--- a/crates/ty_python_semantic/src/types/infer/builder/binary_expressions.rs
+++ b/crates/ty_python_semantic/src/types/infer/builder/binary_expressions.rs
@@ -15,6 +15,22 @@ use ruff_python_ast::PythonVersion;
 use crate::Program;

 impl<'db> TypeInferenceBuilder<'db, '_> {
+    fn infer_binary_operand_expression(
+        &mut self,
+        operand: &ast::Expr,
+        op: ast::Operator,
+        tcx: TypeContext<'db>,
+    ) -> Type<'db> {
+        let operand_tcx = match (op, operand) {
+            // `list.__mul__(int)` preserves the element type, so repeating a list expression can
+            // reuse the outer expected list type during inference.
+            (ast::Operator::Mult, ast::Expr::List(_) | ast::Expr::ListComp(_)) => tcx,
+            _ => TypeContext::default(),
+        };
+
+        self.infer_expression(operand, operand_tcx)
+    }
+
     pub(super) fn infer_binary_expression(
         &mut self,
         binary: &ast::ExprBinOp,
@@ -32,8 +48,8 @@ impl<'db> TypeInferenceBuilder<'db, '_> {
             node_index: _,
         } = binary;

-        let left_ty = self.infer_expression(left, TypeContext::default());
-        let right_ty = self.infer_expression(right, TypeContext::default());
+        let left_ty = self.infer_binary_operand_expression(left, *op, tcx);
+        let right_ty = self.infer_binary_operand_expression(right, *op, tcx);

         self.infer_binary_expression_type(binary.into(), false, left_ty, right_ty, *op)
```

**Comment by oconnor663** [comment]
https://github.com/astral-sh/ruff/pull/23790 already fixed one of the diagnostics in the original example by having `[None] * len(x)` infer as `[None | Unknown]`, but https://github.com/astral-sh/ruff/pull/24197 landed just now to make it infer as `[int | None]`.
